### PR TITLE
[Dubbo-7367]fix too many instance bean created

### PR DESCRIPTION
**Pull request by zhangyz-hd**
## What is the purpose of the change

初步修复 #7367 的问题
当多处用注解订阅同一个服务，并使用parameters时，如
```
@DubboReference(version = "2.0", check = false, parameters = {"a", "1"})
```
实际会产生多个Reference示例

## Brief changelog

1，修复generateReferenceBeanName过程中未处理Array类型的entry问题
2，如果产生多个相同“INTERFACE:VERSION:GROUP”的Reference实例，则在启动日志中增加WARN信息，帮助用户分析
日志示例
```
[25/03/21 07:18:23:790 CST] main  WARN annotation.ReferenceAnnotationBeanPostProcessor:  [DUBBO] ServiceBean:org.apache.dubbo.demo.DemoServiceA1:2.0:Group1 has 2 reference instances, there are: @Reference(check=false,group=Group1,parameters=[a=1,b=2],version=2.0) org.apache.dubbo.demo.DemoServiceA1, @Reference(check=false,group=Group1,parameters=[a=2,b=1],version=2.0) org.apache.dubbo.demo.DemoServiceA1, dubbo version: , current host: 192.168.1.178

```

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [ ] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [ ] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java [hunk 1]
```diff
@@ -212,9 +228,26 @@ private String generateReferenceBeanName(AnnotationAttributes attributes, Class<
         if (!attributes.isEmpty()) {
             beanNameBuilder.append('(');
             for (Map.Entry<String, Object> entry : attributes.entrySet()) {
+                String value;
+                if (entry.getValue().getClass().isArray()) {
+                    String[] entryValues = (String[]) entry.getValue();
+                    if ("parameters".equals(entry.getKey())) {
+                        // parameters spec is {key1,value1,key2,value2}
+                        ArrayList<String> kvList = new ArrayList<>();
+                        for (int i = 0; i < entryValues.length / 2 * 2; i = i + 2) {
```
Comment (zonghaishang) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java:237]: ` i < entryValues.length / 2 * 2` changed to  `i < entryValues.length`.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java [hunk 2]
```diff
@@ -344,4 +377,14 @@ public void destroy() throws Exception {
         this.injectedFieldReferenceBeanCache.clear();
         this.injectedMethodReferenceBeanCache.clear();
     }
+
+    @Override
+    public void onApplicationEvent(ApplicationEvent event) {
+        if (event instanceof ContextRefreshedEvent) {
+            referencedBeanNameIdx.entrySet().stream().filter(e -> e.getValue().size() > 1).forEach(e -> {
```
Comment (zonghaishang) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java:384]: release referencedBeanNameIdx resource after app started.

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java (hunk 1 above)
Comment (zhangyz-hd) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java:237]: It's for protect mistake useage like {key1,value1,key2,value2,key3}

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java (hunk 2 above)
Comment (zhangyz-hd) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java:384]: OK

**Review by zonghaishang — APPROVED** [review]

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java [hunk 3]
```diff
@@ -61,8 +61,9 @@
 
     private final Set<String> serviceNames;
     private final ServiceDiscovery serviceDiscovery;
+    private final String registryId;
     private URL url;
-    private Map<String, NotifyListener> listeners;
+    private Map<String, List<NotifyListener>> listeners;
```
Comment (zonghaishang) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java:66]: List<NotifyListener>改成Set<NotifyListener>

**Review by zonghaishang — CHANGES_REQUESTED** [review]

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java (hunk 3 above)
Comment (zhangyz-hd) [reply dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/client/event/listener/ServiceInstancesChangedListener.java]: OK~

File: dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java [hunk 4]
```diff
@@ -211,10 +229,19 @@ private String generateReferenceBeanName(AnnotationAttributes attributes, Class<
...
-                beanNameBuilder.append(entry.getKey())
+            //sort attributes keys
+            List<String> sortedAttrKeys = new ArrayList<>(attributes.keySet());
+            Collections.sort(sortedAttrKeys);
+            for (String key : sortedAttrKeys) {
+                Object value = attributes.get(key);
+                //handle method array, generic array
+                if (value != null && value.getClass().isArray()) {
+                    Object[] array = ObjectUtils.toObjectArray(value);
+                    value = Arrays.toString(array);
```
Comment (kylixs) [inline dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java:240]: It seems that `@Method`'s `@Argument` array is not handled properly.
```
@DubboReference(methods = @Method(name = "sayHello", timeout = 100, arguments = @Argument(index = 0, callback = true)))
```

Some code snippets:

```
for (String key : sortedAttrKeys) {
      Object value = attributes.get(key);
      value = convertAttribute(value);
      beanNameBuilder.append(key)
              .append('=')
              .append(value)
              .append(',');
}
```

```
    private Object convertAttribute(Object obj) {
        if (obj == null) {
            return null;
        }
        if (obj instanceof Annotation) {
            AnnotationAttributes attributes = AnnotationUtils.getAnnotationAttributes((Annotation) obj, true);
            for (Map.Entry<String, Object> entry : attributes.entrySet()) {
                entry.setValue(convertAttribute(entry.getValue()));
            }
            return attributes;
        } else if (obj.getClass().isArray()) {
            Object[] array = ObjectUtils.toObjectArray(obj);
            Object[] newArray = new Object[array.length];
            for (int i = 0; i < array.length; i++) {
                newArray[i] = convertAttribute(array[i]);
            }
            return Arrays.toString(newArray);
        } else {
            return obj;
        }
    }
```

Comment (zhangyz-hd) [reply dubbo-config/dubbo-config-spring/src/main/java/org/apache/dubbo/config/spring/beans/factory/annotation/ReferenceAnnotationBeanPostProcessor.java:240]: Here will handle arrays well. For Example with annotation below:
```
    @DubboReference(version = "2.0", check = false, parameters = {"a", "1"},
            methods = {
                    @Method(name = "sayHello", timeout = 1000, arguments = {
                            @Argument(callback = true, type = "String"),
                            @Argument(callback = false, type = "int")}
                    ),
                    @Method(name = "test", timeout = 3000)
            }
    )
```
will got the follow string 
```
@Reference(check=false,methods=[@org.apache.dubbo.config.annotation.Method(isReturn=true, cache=, deprecated=false, loadbalance=, actives=0, sent=true, timeout=1000, onreturn=, retries=-1, async=false, oninvoke=, sticky=false, arguments=[@org.apache.dubbo.config.annotation.Argument(type=String, callback=true, index=-1), @org.apache.dubbo.config.annotation.Argument(type=int, callback=false, index=-1)], onthrow=, executes=0, merger=, validation=, name=sayHello), @org.apache.dubbo.config.annotation.Method(isReturn=true, cache=, deprecated=false, loadbalance=, actives=0, sent=true, timeout=3000, onreturn=, retries=-1, async=false, oninvoke=, sticky=false, arguments=[], onthrow=, executes=0, merger=, validation=, name=test)],parameters=[a, 1],version=2.0) org.apache.dubbo.demo.DemoServiceA1
```

**Review by zonghaishang — APPROVED** [review]

**Comment by kylixs** [comment]
LGTM

**Pull request merged**